File: go/p5_mat_lightsout.go
package main

import (
    "g2d"
    "math/rand"
)

// <https://en.wikipedia.org/wiki/Lights_Out_(game)>
type LightsOut struct {
    w, h int
    board      []bool
}

func NewLightsOut(w, h, level int) *LightsOut {
    g := &LightsOut{w, h, make([]bool, w*h)}
    for i := 0; i < level; i++ {
        g.PlayAt(rand.Intn(w), rand.Intn(h))
    }
    return g
}

func (g *LightsOut) Cols() int {
    return g.w
}

func (g *LightsOut) Rows() int {
    return g.h
}

func (g *LightsOut) Finished() bool {
    for y := 0; y < g.h; y++ {
        for x := 0; x < g.w; x++ {
            if g.board[y*g.w+x] {
                return false
            }
        }
    }
    return true
}

func (g *LightsOut) PlayAt(x, y int) {
    // Place (or remove) a light at cell (x, y)
    if 0 <= x && x < g.w && 0 <= y && y < g.h {
        for _, d := range [][]int{{0, 0}, {0, -1}, {1, 0},
            {0, 1}, {-1, 0}} {
            x1, y1 := x+d[0], y+d[1]
            if 0 <= x1 && x1 < g.w && 0 <= y1 && y1 < g.h {
                g.board[y1*g.w+x1] = !g.board[y1*g.w+x1]
            }
        }
    }
}

func (g *LightsOut) FlagAt(x, y int) {
}

func (g *LightsOut) ValueAt(x, y int) string {
    if 0 <= x && x < g.w && 0 <= y && 0 <= y && y < g.h && g.board[y*g.w+x] {
        return "@"
    }
    return "-"
}

func (g *LightsOut) Message() string {
    return "Puzzle solved!"
}

func main() {
    game := NewLightsOut(5, 5, 4)
    g2d.GuiPlay(game)
    //boardgame.ConsolePlay(game)
}

File: examples/p5_mat_lightsout.py
#!/usr/bin/env python3
'''
@author  Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

from boardgame import BoardGame, console_play
from boardgamegui import gui_play
from random import randrange

class LightsOut(BoardGame):
    '''https://en.wikipedia.org/wiki/Lights_Out_(game)'''

    def __init__(self, w=5, h=5, level=4):
        self._w, self._h = w, h
        self._board = [False] * (w*h)
        for _ in range(level):
            self.play_at(randrange(w), randrange(h))

    def cols(self) -> int:
        return self._w

    def rows(self) -> int:
        return self._h

    def finished(self) -> bool:
        # https://docs.python.org/3/library/functions.html#any
        return not any(self._board)

    def play_at(self, x: int, y: int):
        '''Place (or remove) a light at cell (x, y)'''
        w, h = self._w, self._h
        if 0 <= x < w and 0 <= y < h:
            for dx, dy in ((0, 0), (0, -1), (1, 0),
                           (0, 1), (-1, 0)):
                x1, y1 = x + dx, y + dy
                if 0 <= x1 < w and 0 <= y1 < h:
                    self._board[y1*w+x1] = not self._board[y1*w+x1]

    def flag_at(self, x: int, y: int):
        pass

    def value_at(self, x: int, y: int) -> str:
        w, h = self._w, self._h
        if 0 <= x < w and 0 <= y < h and self._board[y*w+x]:
            return '@'
        return '-'

    def message(self) -> str:
        return "Puzzle solved!"


def main():
    game = LightsOut()
    gui_play(game)
    ##console_play(game)

main()
